cli/cli-extensions.md

```diff
@@ -0,0 +1,195 @@
+# CLI Extensions
+
+## Overview
+
+CLI extensions are separate executables that appear as part of the `temporal` CLI. This is similar to
+[`kubectl` plugins](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) and `git` extensions/commands.
+
+## Usage Examples
+
+### Custom Authenticator
+
+A custom extension can allow
+
+    temporal mycompany login
+
+To set some environment config settings on disk so that successive calls to
+
+    temporal workflow start --type foo --task-queue some-tq --workflow-id foo-id
+
+automatically work.
+
+### Shortcuts for Common Tasks
+
+A custom extension can add something like
+
+    temporal mycompany do-company-thing
+
+And that can do anything programmatically and also get the benefit of having access to the Temporal client. So it could
+access some company resource (e.g. a database) and send an update to a workflow with some value.
+
+### Contributions
+
+Custom extensions allow an ecosystem of CLI extensions to form that do not require Temporal oversight. For example
+
+    temporal workflow show-diagram --workflow-id foo-id
+
+Could be an extension that downloads the workflow history and makes a nice visualization of it.
+
+### Cloud CLI
+
+A Temporal Cloud extension will be available (and likely released alongside the traditional CLI) that will allow
+cloud-specific operations like:
+
+    temporal cloud namespace list
+
+This can share authentication with the CLI in many ways. It can even support login, so the following two commands could
+work well together:
+
+    temporal cloud login --profile cloud
+    temporal workflow start --profile cloud --type foo --task-queue some-tq --workflow-id foo-id
+
+These are just usage examples for the purposes of explaining extensions, this may not be what `temporal cloud` extension
+looks like when built.
+
+## Runtime Behavior
+
+### Lookup
+
+Extensions are any executables on the `PATH` with `temporal-` prefix. Extension executable names use `-` for each
+subcommand and `_` for each dash in the command (but still can be called with an underscore). This is similar to
+`kubectl` behavior.
```

Comment:
While there is precedent with kubectl and git, I wonder whether we should choose a slightly less common prefix such as `temporalcli-` or `temporal-cli-` or `-temporal-`. I've surveyed one temporal user, and they already have the following executables named `temporal-` (found by issuing `temporal-<TAB>`):

```
cli(main) temporal-
temporal-cancel-all
temporal-delete-all
temporal-generate-help-output
temporal-github-actions-delete-runs
temporal-otel-collector
temporal-python-release.sh
temporal-python-test-failures
temporal-samples-traffic
temporal-server
temporal-server-test-failures
temporal-slack-search
temporal-terminate-all
temporal-typescript-link
temporal-workflow-from-url
temporal-workflow-list-ids
temporal-workflow-start
```

Reply:
I expect the dashes to be like spaces on a command line. Looking at that list, I'd expect those to be CLI extensions same as I would for other extensible CLIs. All of those all now magically work with spaces instead of dashes with the exception of `temporal-server` and `temporal-workflow-start`. I don't think we should alter the simplicity of extension names based on this user's list.